3 1

agent96 opened the pull request:
Changes to the SDMX 3.1 Schema undoing the work on reversing content constraints, including the additional related modifications to the Constraint model which made the model more reusable.

Changes:
1. Namespaces change to 3_1
2. SDMXStructure.xsd Added AvailabilityConstraints
3. SDMXStructureBase.xsd ItemType parent ID changes from nestedID to IDType
4. SDMXStructureCategorisation.xsd version prohibited (fixed to 1.0)
5. SDMXStructureCodelist.xsd parent changes from IDType to SingleNCNameIDType 
6. SDMXStructureConcept.xsd parent changes from IDType to SingleNCNameIDType 
7. SDMXStructureConstraint.xsd 
    1.  DataConstraintBaseType remove ReleaseCalendar and Role
    * MetadataConstraintBaseType remove ReleaseCalendar and Role
    * ConstraintAttachmentType simplify by removing URLs
    * Added AvailabilityConstraintType
    * DataKeyValueType support unbounded Value in member selection type (FREQ=A or M or Q  and REF_AREA=DE and CURR=EUR)
* SDMXStructureDataStructure.xsd 
    * Added evolvingStructure property to the DataStructureType
* SDMXStructureDataflow.xsd added optional DimensionConstraint property

Commit:
Re-introduce metadata constraint

Commit:
update to v3_1

Commit:
first commit

Commit:
namespace change was 3_1_0 now 3_1

Commit:
not used

Commit:
constraint attachment re-introduced

Commit:
version fixed and prohibited

Commit:
reintroduce MetadataTargetRegion

Commit:
reintroduce MetadataConstraints

Commit:
re-introduce metadata constraint and constraint attachment

Commit:
maxOccurs="1" on availability constraint

Commit:
remove reference from constrainable to constraints

Commit:
first commit

Commit:
no longer used

Commit:
remove dimension constraint

Commit:
remove DimensionConstraints

Commit:
remove DimensionConstraintBaseType

Commit:
added DimensionConstraintType

Commit:
reverted change to remove KeyValue on cube region

Commit:
update documentation on DimensionConstraint

Commit:
added evolvingStructure property to DSD

Commit:
change Component to KeyValue

Commit:
Change Component to KeyValue

Commit:
remove defaultValue

Commit:
bug fixes and typos

Commit:
added default as false

Commit:
https://github.com/sdmx-twg/sdmx-ml/issues/33

Commit:
fix categorisation URN to 1.0

Commit:
removed DataConstraintAttachmentType and
MetadataConstraintAttachmentType (a hangover from the constraint
reversal design)

Commit:
bug fix on semantic versioning reference pattern
https://github.com/sdmx-twg/sdmx-ml/issues/42

Commit:
move XML comment to XSD annotation

Commit:
updated Copyright to 2025

aboddie committed:
Merge pull request #49 from sdmx-twg/3_1_issue42

bug fix on semantic versioning reference pattern

aboddie committed:
Remove reference to ReportingYearStartDay

dosse committed:
Aditional corrections

aboddie committed:
Update indentation in SDMXStructureConstraint.xsd

Commit:
formatting of xs:schema attributes, added version=3.1.0

Commit:
formatting of mes:Structure and added absolute schema location

aboddie committed:
Merge pull request #50 from aboddie/3_1

Remove reference to ReportingYearStartDay

aboddie committed:
Remove <p> from with documentation tags

aboddie committed:
Fix spelling mistakes

aboddie committed:
revert removing ReportingYearStartDay which seems to be used in the API.

agent96 committed:
remove version =3.1.0

agent96 committed:
updated samples to use xml.sdmx.org (previously xmlschema.sdmx.org)

agent96 committed:
added schema location to readme

Commit:
updated documents for 3.1

Commit:
Merge branch '3_1' of git@github.com:sdmx-twg/sdmx-ml.git into 3_1

Commit:
Merge branch '3_1' of git@github.com:sdmx-twg/sdmx-ml.git into 3_1

dosse committed:
Merge pull request #51 from sdmx-twg/3_1_schemalocation

3 1 schemalocation

dosse committed:
Merge branch 'develop' into 3_1

dosse commented:
@agent96 Before merging, I still wanted to address:
- https://github.com/sdmx-twg/sdmx-ml/issues/31
- https://github.com/sdmx-twg/sdmx-ml/issues/32

Could you give me until the end of the week please?

agent96 commented:
yes that's fine, thanks @dosse

dosse committed:
https://github.com/sdmx-twg/sdmx-ml/issues/31, https://github.com/sdmx-twg/sdmx-ml/issues/32

dosse commented:
Thanks, @agent96! I believe I finished, but I'm not fully sure if anything else needs to be changed in the xml schemas to allow for special dimension values "~" in data messages (for higher-level data-related ref. metadata) and for special intentionally missing measure and attribute values (NaN, "#N/A"). What do you think?

PS. I'm travelling and my network connection is bad. I couldn't get an XML validator to work with the new schemas and examples to do the related test. I try later today or tomorrow, unless you could already confirm.

dosse reviewed (dismissed):
Ok for me.

agent96 commented:
Thanks @dosse - for paranoia I will recheck the samples.  For the special character, I believe this is only required in the generated schema files (based on the DSD) rather then the fixed schemas we provide here.

Commit:
support for missing types

agent96 commented:
@dosse I have updated the SDMXCommon.xsd with a MissingType (~) and added support for common types which allow missing i.e IntOrMissing which is a union of xs:int and MissingType.  I have also updated section 3A by adding section 3.3.7 "Representation with Missing Value Support" which describes how structure specific schemas have to allow missing values in Dimensions when the target is a DSD with the evolving structure property set to true.

I have also checked all the samples (there were a number of errors in the namespace declarations which have been fixed).  

The other issue was introduced on the recent Component2Attribute merge into develop, it was not syntactically correct and caused the schema to fail - the problem is the RegionType can not be restricted by renaming Component to Attribute, it would need a new type - I suggest we simply revert this and keep everything as it is to avoid any complications.  Changes have been made in this branch to ensure everything now validates.

Commit:
bug fix RegionType-Attribute becomes Component

Commit:
bug fix on namespace

Commit:
bug fix on sample (typo in namespace)

dosse committed:
https://github.com/sdmx-twg/sdmx-ml/issues/53

dosse commented:
Thanks a lot for the verifications and corrections! All changes seem ok to me. I just realise that I still need to check some things in XML, like the `isPartialLanguage` implementation.

dosse committed:
https://github.com/sdmx-twg/sdmx-ml/issues/53

dosse commented:
@agent96 

I updated the docs for COMMON, DATA and STRUCTURES for the `isPartialLanguage` implementation. Since it was agreed for Metadatasets, a maintainable, I have added it to all maintainables.  By any chance, would you have a Metadataset example, which we could add to the GitHub?

About the "missing value" type (https://github.com/sdmx-twg/sdmx-ml/pull/48/commits/1753a14a01c9d4ee24fa4b6b5ac066267cf5cde1) for dimension values, to differentiate it from the intentionally missing measure and attribute values, which use `NaN` and `#N/A`, I changed the textual description in the docs from "missing" to "irrelevant".  But I didn't change the type name "common:MissingType".  You are right that for measures and attributes not attached to a dimension if the DSD allows for an ‘evolving structure’, this cannot be the time dimension. However, for irrelevant dimensions in data-related higher-level (i.e., attached to dataflow or partial list of Dimensions) reference metadata attributes that don’t have a fixed pre-defined attachment, also the time dimension can be irrelevant.  Does this still need to be reflected in the schemas?

Did you also test if the new schemas allow using `NaN` or `#N/A` for non-dimension component values? Should we add an example?

agent96 commented:
@dosse thanks for the review. Very good points. Section 3A Part IV Data contains the documentation on schema generation for data, the missing types I added to the SDMX schema are prebuilt unions for convenience.  I could also add types including #N/A to be consistent.

The main change thought would be to the documentation as it will defines how structure specific schema are generated, this will need to include support for #N/A.  I will take this action.  

I believe there are the following changes required:
1.  #N/A is always a valid value for observation values and attribute when the data type is not float or double data type.  When the data type is float or double NaN is already supported.  
2. ~ is a valid value for metadata attributes in all case
3. change the wording in the document and schema so we use consistent terminology. To confirm: 
3a. NaN and #N/A is used to represent intentionally missing
3b ~ is used to represent irrelevant

dosse commented:
@agent96 Thanks a lot!

For the question on requied changes: I understood that:
- "~" is only for values of dimensions, never for values of measures or (metadata) attributes. It indicates "not applicable" dimension, for instance for DSD-extensions or for data-related ref. metedata not attached to certain dimensions.
- `NaN` and `#N/A` are only for values of measures and (metadata) attributes, never for values of dimensions. It indicates "intentionally missing/not available" values for those non-dimension components. `NaN` is only for values of type float or double. All other types use `#N/A`.

Commit:
support for #N/A

Commit:
Merge branch '3_1' of git@github.com:sdmx-twg/sdmx-ml.git into 3_1

Commit:
report not applicable

agent96 commented:
@dosse I have updated the schema with the two base types (Missing and NotApplicable) and updated the section 3A data document to describe both.  The two data schemas for ECB EXR have been updated so that all attributes support #N/A and I have updated both dataset samples to include OBS_STATUS="#N/A".  It would be useful to get some other examples in for horizontally complex DSD with not applicable values, but we either need a good real world example or a very simple made up example. Perhaps just demonstrating missing values is enough as the two operate in a very similar way.

Commit:
added missing value

dosse commented:
@agent96 Very good! Thanks a lot. I guess we can add more examples for horizontally complex DSD with not applicable values and for ref. metadatasets through patch changes in the future, if we can't wait any longer with the release?

dosse committed:
Update SDMXCommon.xsd

Correction of doc text

dosse committed:
Update SDMXCommon.xsd

Corrected #N/A to ~ for NotApplicableType

dosse commented:
@agent96 When updating the Word documents, you seem to not use the always current last version in the branch, loosing the changes I made. I will reapply my changes.

agent96 commented:
@dosse sorry about that - I don't know how that happened.

dosse committed:
doc recovery/update and some markdown

dosse committed:
one more markdown

dosse committed:
remaining markdowns

dosse reviewed (approved):
All fine for me!  
Updated the docx to latest version and also translated to markdown.
